pkg/alertmanager/multitenant_test.go

```diff
@@ -1161,10 +1159,22 @@ func TestAlertmanager_StateReplicationWithSharding(t *testing.T) {
 
 			// With sharding enabled, we propagate messages over gRPC instead of using a gossip over TCP.
 			// 1. First, get a random multitenant instance
-			multitenantAM := instances[rand.Intn(len(instances))]
+			//    We must pick an instance which actually has a user configured.
+			var multitenantAM *MultitenantAlertmanager
```

Comment:
I'm not sure if this holds true... Perhaps @pstibrany knows more?

As part of the test setup I'm doing two things:

1. Make sure I set the polling interval to a high number - in this case, an hour to avoid syncing configuration in the middle of the test.
2. Make sure the ring has settled by waiting until all the instances are ready.

With those two premises, all the tenants should have been assigned and ready at this point - because we don't change the state until we've done that sync.

Now, I _think_ there might be a case here not taken into account by lines 1144 - 1148, let's say that the first instance was set to ready but as we moved into the second one that one also re-sharded we could end up in a state where the third one is ready but the other two have changed their status.

Reply:
Update, discussed this offline. The case that fails is RF=3, 10 instances. With 12 configs that are a total of 36 configs across 10 replicas, with the way the numbers line up there could be a replica that ends up with 0.

Reply:
Just for reference, logs from one of the tests:
```
out.log:85:level=debug instance=alertmanager-1 component=MultiTenantAlertmanager msg="adding configurations" num_configs=4
out.log:135:level=debug instance=alertmanager-2 component=MultiTenantAlertmanager msg="adding configurations" num_configs=7
out.log:220:level=debug instance=alertmanager-3 component=MultiTenantAlertmanager msg="adding configurations" num_configs=4
out.log:267:level=debug instance=alertmanager-4 component=MultiTenantAlertmanager msg="adding configurations" num_configs=3
out.log:303:level=debug instance=alertmanager-5 component=MultiTenantAlertmanager msg="adding configurations" num_configs=6
out.log:371:level=debug instance=alertmanager-6 component=MultiTenantAlertmanager msg="adding configurations" num_configs=2
out.log:395:level=debug instance=alertmanager-7 component=MultiTenantAlertmanager msg="adding configurations" num_configs=1
out.log:408:level=debug instance=alertmanager-8 component=MultiTenantAlertmanager msg="adding configurations" num_configs=0
out.log:410:level=debug instance=alertmanager-9 component=MultiTenantAlertmanager msg="adding configurations" num_configs=5
out.log:467:level=debug instance=alertmanager-10 component=MultiTenantAlertmanager msg="adding configurations" num_configs=4
```